Blogpost (#8)
marcgs/lob-agent-evaluation: blog_post.md

> # Evaluating LLM-Powered Chatbots in Enterprise Workflows

Reviewer: Let's move this document to `/docs/evaluation`

## Introduction

The landscape of conversational AI has undergone a seismic shift with the emergence of large language models (LLMs). These sophisticated systems have transformed chatbots from rigid, rule-based interactions to dynamic, seemingly intelligent conversations. But with this transformation comes a fundamental challenge: how do we effectively measure the performance of these systems, especially in business-critical environments?

> ![LOB Chatbot Demo](./chatbot_demo.gif)

Reviewer: I would vote for moving the animated gif at the end of the "Introduction" section. It gives the reader more context to digest, before having a visualization.

*If you would like to better understand the role of "agents" in conversational AI, I encourage you to check out [building effective agents](https://www.anthropic.com/engineering/building-effective-agents) blog post by Anthropic.*

### The Non-Deterministic Evaluation Problem

LLM-based chatbots introduce a fundamental challenge to evaluation: non-determinism. Unlike rule-based systems that produce identical outputs for identical inputs, LLMs incorporate elements of randomness and contextual interpretation that can result in varied—yet equally valid—responses to the same query.

This non-determinism creates several evaluation hurdles:

- **Reproducibility issues:** Test cases may pass or fail inconsistently
- **Ground truth ambiguity:** Multiple response paths may be equally correct

> - **Function calling reliability:** The same user intent might be mapped to different function calls on different runs

Reviewer: I don't think we are addressing this challenge in our sample. No action required, just wanted to raise this to discuss.

Author: Not all of it, yes. A lot of it indirectly by providing a way of running chat simulations at scale. In any case in this section we don't claim we have solution for all of these. We just state that these are the challenges due to non-deterministic nature of LLM agents. So all in all, good that you're raising this but I think it's OK to have it.

## LOB Chatbot Evaluation Framework

Our team have recently worked with one of our customers on developing a LOB chatbot that was supposed to replace an internal, legacy tool for change management in their product development process. The tool was hard to maintain, had a very complex UI with a lot of business rules and restrictions, and it also required a lot of expert knowledge from users. Their leadership believed LLM-powered chatbots are the answer for these challenges and they wanted to build one with us. Our task was to prove feasibility and reliability of such solution when deployed at scale. That's where we started considering all of the forementioned challenges in evaluating such systems. Our primary goal was to run evaluations which would prove overall reliability and provide stakeholders with enough confidence in this solution. With that in mind we designed and implemented an evalution framework for LLM-powered, LOB chatbots.

> TODO: Update evaluation framework diagram to make it super general. Add error analysis.

Reviewer: Reminder for the TODO. Would be great to improve the resolution of the visual

Your goal is to complete a specific task by conversing naturally with the assistant.

Behave like a non-technical user who understands the task, but not the internal workings of the system.

You have access to some business data relevant to your task. Use it when appropriate during the conversation.

Follow these rules:
- Act as a user, not an assitant.
- Never correct the assistant or point out mistakes.
- You are not allowed to change the inputs proposed by the assistant.
- Stay focused on the task but allow for slight variability in how you express yourself.
- Use the business data as needed, but do not mention that it was "given" to you.
- Do not modify the business data you are given.
- Speak naturally, as if you are recalling or referencing information you know.
- Once your task is completed, you must end the conversation by saying "the session is finished".
- Your goal is achieved when the assistant has completed the task and you have confirmed it.

Here is the business data you can use during the conversation:
{business_data}

Here is your task:
{user_instructions}

Begin the conversation. Respond as the user.
"""
```

> To ensure realistic interactions, the User Agent is prompted to simulate non-technical users who understand the task but not the internal workings of the system. However, the simulation framework we've built is flexible enough to support creation of test cases which can take into account different user personalities and interaction styles. This approach helps identify potential usability issues and ensures the chatbot can handle diverse user expressions effectively. For example, in real world we might find power users who will be very concise and efficient in their interactions with the chatbot - with our approach we can easily test it by providing additional instructions to the User Agent:

Reviewer: ```suggestion
To ensure realistic interactions, the User Agent is prompted to simulate non-technical users who understand the task but not the internal workings of the system. However, the simulation framework we've built is flexible enough to support creation of test cases which can take into account different user personalities and interaction styles. This approach helps identify potential usability issues and ensures the chatbot can handle diverse user expressions effectively. For example, in real-world scenarios, we may encounter power users who interact with the chatbot in a concise and efficient manner, as well as novice users whose interactions are less efficient and prone to mistakes. With our approach we can easily test it by providing additional instructions to the User Agent:
```

```json
...
"scenario_name": "cmd_create_ticket_and_action_item",
"user_instructions": "Create a new support ticket by manually providing the data and create a new action item for it. Don't use full sentences but only command-like terms 'search ...', 'new action ...', etc. If presented with a numerical list of options, use the number of the option to select it.",
...
```

> 
> ```json
> Conversation snippet:

Reviewer: ```suggestion

Conversation snippet:
```json
```

{
  "role": "assistant",
  "content": "Hello! How can I assist you today? Here are your options:\n\n1. Create a new support ticket.  \n2. Update an existing support ticket.  \n3. Create a new action item for an existing ticket.  \n4. Update an existing action item.  \n5. Search for historical tickets.  \n\nPlease let me know how you'd like to proceed!",
  "name": "SupportTicketAgent"
},
{
  "role": "user",
  "content": "1",
  "name": "UserAgent"
},
{
  "role": "assistant",
  "content": "Great! You want to create a new support ticket. How would you like to proceed?\n\n1. Manually provide the required details for the support ticket.  \n2. Use information from a similar previous ticket as a template.  \n\nLet me know your preference!",
  "name": "SupportTicketAgent"
},
{
  "role": "user",
  "content": "1",
  "name": "UserAgent"
},
{
  "role": "assistant",

>   "content": "Got it! Please provide the following details to create the support ticket:\n\n1. **Title**: A concise and descriptive

Reviewer: Let's try to have valid json:
```suggestion
  "content": "Got it! Please provide the following details to create the support ticket:\n\n1. **Title**: A concise and descriptive ..."
```

          "department_code": "{ticket.department_code}",
          "priority": "{ticket.priority}",
          "workflow_type": "{ticket.workflow_type}",
          "description": "{ticket.description}",
          "expected_outcome": "{ticket.expected_outcome}",
          "customer_visible": "{ticket.customer_visible}"
        }
      },
      {
        "functionName": "ActionItemPlugin-create_action_item",
        "arguments": {
          "title": "{action.title}",
          "assignee": "{action.assignee}",
          "due_date": "{action.due_date}"
        }
      }
    ]
}
```

### Simulation Outputs

Throughout the simulated conversation, the framework automatically captures all exchanged messages and every function call made by the agent, including the function name and all arguments provided. This structured record of function calls is essential for evaluation: it allows us to directly compare the chatbot's actions against the expected function calls for each scenario, measuring not just whether the right functions were called, but also whether the correct parameters were supplied and the business process was followed as intended.

Simulation outputs:

> ![Simulation Outputs](./sim_outputs.png)

Reviewer: I think providing all simulation outputs as markdown json fenced blocks would be more in line with the rest of the document and IMO more readable as well.

Simulated conversation history:

![Chat History](./chat_history.png)

Actual vs expected function calls - used to calculate various metrics and overall reliability score of the LOB agent:

![Function Calls](./function_calls.png)

### Evaluation, Metrics, and Error Analysis

With conversation history, actual and expected function call data in hand, we can use the framework to compute a suite of evaluation metrics, which include:

> - **Function Call Name Precision and Recall**  

Reviewer: Should we reorganize the metrics by:
- Primary metrics:
  - Function Call Name Recall
  - Function Call Parameter Recall
 - Auxiliary metrics:
   - Function Call Name Precision
   - Function Call Parameter Precision
   - Reliability Score

Author: I think we don't need to do this. Current structure makes more from metrics perspective and in Reliability Score we explain recall is our primary metric

  Measures the accuracy of function calling, in terms of function names.  
  For recall we check how many of expected functions were correctly called by LOB agent, and for precision we measure if any unexpected functions were called during simulation (LLMs can halucinate or simply misspell function names).
- **Function Call Argument Precision and Recall**  

>   Measures the accuracy of function calls in terms of function parameters.  

Reviewer: ```suggestion
  Measures the accuracy of function calls, in terms of function parameters.  
```
